go/src/koding/kites/kloud/provider/google/google.go

```diff
@@ -0,0 +1,164 @@
+package google
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+
+	"koding/kites/kloud/stack"
+	"koding/kites/kloud/stack/provider"
+
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2/google"
+	compute "google.golang.org/api/compute/v1"
+)
+
+var p = &provider.Provider{
+	Name:         "google",
+	ResourceName: "compute_instance",
+	Machine:      newMachine,
+	Stack:        newStack,
+	Schema: &provider.Schema{
+		NewCredential: newCredential,
+		NewBootstrap:  newBootstrap,
+		NewMetadata:   newMetadata,
+	},
+}
+
+func init() {
+	provider.Register(p)
+}
+
+// Region represents google's geographical region code.
+type Region string
+
+var regions = []Region{
+	"asia-east1",
+	"europe-west1",
+	"us-central1",
+	"us-east1",
+	"us-west1",
+}
+
+// Enum returns all available regions for "google" provider.
+func (Region) Enum() (rs []interface{}) {
+	for _, region := range regions {
+		rs = append(rs, region)
+	}
+	return rs
+}
+
+// Valid checks if stored region code is available in GCP.
+func (r Region) Valid() error {
+	if r == "" {
+		return fmt.Errorf("region name is not set")
+	}
+
+	for _, region := range regions {
+		if r == region {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("unknown region name: %v", r)
+}
+
+// Cred represents jCredentialDatas.meta for "google" provider.
+type Cred struct {
+	Credentials string `json:"credentials" bson:"credentials" hcl:"credentials" kloud:",secret"`
+	Project     string `json:"project" bson:"project" hcl:"project" kloud:",secret"`
+	Region      Region `json:"region" bson:"region" hcl:"region"`
+}
+
+var _ stack.Validator = (*Cred)(nil)
+
+func newCredential() interface{} {
+	return &Cred{}
+}
+
+func (c *Cred) Valid() error {
+	if c.Credentials == "" {
+		return errors.New(`cred value for "credentials" is empty`)
+	}
+	if c.Project == "" {
+		return errors.New(`cred value for "project" is empty`)
+	}
+
+	return c.Region.Valid()
+}
+
+func (c *Cred) ComputeService() (*compute.Service, error) {
+	cfg, err := google.JWTConfigFromJSON([]byte(c.Credentials), compute.ComputeScope)
+	if err != nil {
+		return nil, err
+	}
+
+	return compute.New(cfg.Client(context.Background()))
+}
+
+type Bootstrap struct{}
```

Comment:
This should define at least one field that is going to be updated after bootstrap. Empty or nil Bootstrap means no bootstrap, and we do bootstrap for google provider.